### DISCUSSION TITLE
Doing a reduce on arange without materializing the vector

### DISCUSSION THREAD

**Discussion by mgostIH** [jax-ml/jax#10405] [Q&A]
```py
@partial(jax.jit, static_argnums=[0])
def do_funny_thing(N):
    v = jnp.arange(N)
    return jax.lax.reduce(v, 0, lambda x, y: x ^ y, [0])

do_funny_thing(1_000_000_000_000)
```

The following code goes out of memory, it would be lovely if it wouldn't. Is there some known method to avoid materializing an arange, and more generally any kind of vector that is defined in terms of it, like those generated from jax.random? I'm not sure whether it's the reduce actually materializing it in memory, if that's the case what should I do?

**Comment by YouJiacheng** [comment]
You can try `lax.while_loop`. Note this might be slow, you can loop over chunks instead of elements.
alternatively, you can convert [i for i in range(1000)] into [i * 1000 + j for i in range(1000) for j in range(1000)], i.e. a nested scan on two arange.
see https://github.com/google/jax/discussions/10401#discussioncomment-2610609

**Reply by mgostIH** [reply]
Indeed, I thought about the while loop but I'd imagine this would be extremely slow due to its sequential nature, while a lot of the tasks I want to do would be based on reduce.

**Reply by YouJiacheng** [reply]
So you should not process only one element per iteration in while loop!
```python
jax.lax.fori_loop(0, 1_000_0000, lambda i, val: jax.lax.reduce(jnp.arange(i * 1_000_000, (i + 1) * 1_000_000), val, lambda x, y: x ^ y, [0]), 0)
```

**Reply by mgostIH** [reply]
I will consider this as a possible alternative, but I thought that not having to load anything on device and not needing to think about the loop splitting would be a better design for trivially parallelizable tasks.
